Monitoring as the Operator for RHEL by using JMX (#31)
rh-messaging/amq-streams-docs: books/con-important-broker-metrics.adoc

*Kafka server metrics*

The following table shows a selection of metrics that report information about the Kafka server.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected value

¦Messages in per second
m¦kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec
¦The rate at which individual messages are consumed by the broker.
¦Approximately even when compared with the other brokers.

¦Bytes in per second
m¦kafka.server:type=BrokerTopicMetrics,name=BytesInPerSec
¦The rate at which data sent from producers is consumed by the broker.
¦Approximately even when compared with the other brokers.

¦Replication bytes in per second
m¦kafka.server:type=BrokerTopicMetrics,name=ReplicationBytesInPerSec
¦The rate at which data sent from other brokers is consumed by the broker.
¦N/A

¦Bytes out per second
m¦kafka.server:type=BrokerTopicMetrics,name=BytesOutPerSec
¦The rate at which data is fetched and read from the broker by consumers.
¦N/A

¦Replication bytes out per second
m¦kafka.server:type=BrokerTopicMetrics,name=ReplicationBytesOutPerSec

> ¦The rate at which data is sent from the broker to other brokers.

Author: Suggest changing as follows:

> The rate at which data is sent from the broker to other brokers. This metric is useful to monitor if the broker is a leader for a group of partitions.

¦N/A

¦Under-replicated partitions
m¦kafka.server:type=ReplicaManager,name=UnderReplicatedPartitions
¦The number of partitions that have not been fully replicated in the follower replicas. 
¦Zero

¦Under minimum ISR partition count
m¦kafka.server:type=ReplicaManager,name=UnderMinIsrPartitionCount

> ¦The number of partitions under the minimum In-Sync Replica (ISR) count.

Author: Agreed. Suggest changing as follows:

> The number of partitions under the minimum In-Sync Replica (ISR) count. The ISR count indicates the set of replicas that are up-to-date with the leader.

¦Zero

¦Partition count
m¦kafka.server:type=ReplicaManager,name=PartitionCount
¦The number of partitions in the broker.
¦Approximately even when compared with the other brokers.

¦Leader count
m¦kafka.server:type=ReplicaManager,name=LeaderCount

> ¦The number of leader replicas in the broker.

Reviewer: ```suggestion
¦The number of replicas for which this broker is the leader.
```

¦Approximately even when compared with the other brokers.

¦ISR shrinks per second
m¦kafka.server:type=ReplicaManager,name=IsrShrinksPerSec
¦The rate at which the number of ISRs in the broker decreases
¦Zero

¦ISR expands per second
m¦kafka.server:type=ReplicaManager,name=IsrExpandsPerSec
¦The rate at which the number of ISRs in the broker increases.
¦Zero

¦Maximum lag
m¦kafka.server:type=ReplicaFetcherManager,name=MaxLag,clientId=Replica
¦The maximum lag between the time that messages are received by the leader replica and by the follower replicas.
¦Proportional to the maximum batch size of a produce request.

¦Requests in producer purgatory
m¦kafka.server:type=DelayedOperationPurgatory,name=PurgatorySize,delayedOperation=Produce
¦The number of send requests in the producer purgatory.
¦N/A

¦Requests in fetch purgatory
m¦kafka.server:type=DelayedOperationPurgatory,name=PurgatorySize,delayedOperation=Fetch
¦The number of fetch requests in the fetch purgatory.
¦N/A

¦Request handler average idle percent
m¦kafka.server:type=KafkaRequestHandlerPool,name=RequestHandlerAvgIdlePercent

> ¦The average amount of time that the request handler is idle.

Reviewer: If it's an amount of time do we know the units? Milliseconds?

¦A lower value indicates that the workload of the broker is high. 

> ¦Request

Reviewer: ```suggestion
¦Requests exempt from throttling
```

m¦kafka.server:type=Request
¦The number of requests that are exempt from throttling.
¦N/A

¦Zookeeper request latency (milliseconds)
m¦kafka.server:type=ZooKeeperClientMetrics,name=ZooKeeperRequestLatencyMs
¦The latency for ZooKeeper requests from the broker, in milliseconds.
¦N/A

¦Zookeeper session state
m¦kafka.server:type=SessionExpireListener,name=SessionState
¦The status of the broker's connection to Zookeeper.
¦CONNECTED

|===

*Kafka network metrics*

The following table shows a selection of metrics that report information about requests.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected value

¦Requests per second

> m¦kafka.network:type=RequestMetrics,name=RequestsPerSec,request={Produce|FetchConsumer|FetchFollower}

Reviewer: `{Produce|FetchConsumer|FetchFollower}` is neither an MBean ObjectName pattern, nor a regex. I know what you mean, but we need to explain what we mean with the syntax.

¦The total number of requests made for the request type per second.
¦N/A

¦Request size in bytes
m¦kafka.network:type=RequestMetrics,name=RequestBytes,request=([-.\w]+)

> ¦The size of requests made for each request type.

Reviewer: The _average_ size of requests of the given type? You need to make more clear what you mean by "each request type" (i.e. there's a different MBean for each request type, identified by the `request` key property of the MBean name).

¦N/A

¦Temporary memory size in bytes
m¦kafka.network:type=RequestMetrics,name=TemporaryMemoryBytes,request={Produce|Fetch}
¦The amount of temporary memory used for converting message formats and decompressing messages.
¦N/A

¦Message conversions time
m¦kafka.network:type=RequestMetrics,name=MessageConversionsTimeMs,request={Produce|Fetch}
¦Time in milliseconds spent on converting message formats.
¦N/A

¦Total request time in milliseconds
m¦kafka.network:type=RequestMetrics,name=TotalTimeMs,request={Produce|FetchConsumer|FetchFollower}
¦Total time in milliseconds spent processing requests.
¦N/A

¦Request queue time in milliseconds
m¦kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request={Produce|FetchConsumer|FetchFollower}

> ¦The length of time the request spends waiting in the request queue.

Reviewer: "the request" supposes we have a common idea of what request you're referring to. 
```suggestion
The time, in milliseconds, that a request currently spends in the queue for the request type given in the `request` property.
```

¦N/A

> ¦Local time in milliseconds

Reviewer: I think this is a really confusing name. 

```suggestion
¦Leader local processing time
```

Author: I'm trying to name them exactly as they appear in the MBeans. IMO, giving them a new name is more confusing!

Reviewer: Wdym by "as they appear in the MBeans"? IFAIK the MBeans just have an `ObjectName`, which is `kafka.network:type=RequestMetrics,name=LocalTimeMs,request=Produce`, for example. The MBean info for that looks like this in jconsole

![image](https://user-images.githubusercontent.com/879487/48938371-f42a7480-ef08-11e8-986c-ca3932fe1054.png)

There is no "Local time in milliseconds" there.

Author: Sorry, I meant the name: "LocalTimeMs".

Author: My thinking was that it might be difficult for the user to find the metric if we give it a name that's unrelated to the actual MBean. This is an issue with how they've named the MBeans in Kafka.

Reviewer: Ah, OK, I see. I wonder how helpful that really is then, when not all of the metrics have a very good `name` property. We're specifying the whole `ObjectName` anyway, so in a way it's redundant. OTOH it does help highlight what is the name part of the `ObjectName`, so it's not useless.

m¦kafka.network:type=RequestMetrics,name=LocalTimeMs,request={Produce|FetchConsumer|FetchFollower}
¦The time taken for the leader to process the request.
¦N/A

> ¦Remote time in milliseconds

Reviewer: ```suggestion
¦Leader remote processing time
```

m¦kafka.network:type=RequestMetrics,name=RemoteTimeMs,request={Produce|FetchConsumer|FetchFollower}

> ¦The length of time the request waits for the follower.

Reviewer: Is this an average across all the partitions that this broker is leader for?

¦N/A

¦Response queue time in milliseconds
m¦kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request={Produce|FetchConsumer|FetchFollower}

> ¦The length of time the request waits in the response queue.

Reviewer: units

¦N/A

¦Response send time
m¦kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request={Produce|FetchConsumer|FetchFollower}
¦The time taken to send the response.
¦N/A

¦Network processor average idle percent
m¦kafka.network:type=SocketServer,name=NetworkProcessorAvgIdlePercent
¦The average percentage of time that the network processors are idle.
¦Between zero and one.

|===

*Kafka log metrics*

The following table shows a selection of metrics that report information about logging.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected Value

¦Log flush rate and time in milliseconds
m¦kafka.log:type=LogFlushStats,name=LogFlushRateAndTimeMs

> ¦Log flush rate and time.

Reviewer: Here the name is more descriptive then the description!

¦Zero

|===

*Kafka controller metrics*

The following table shows a selection of metrics that report information about the controller of the cluster.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected Value

¦Active controller count
m¦kafka.controller:type=KafkaController,name=ActiveControllerCount
¦The number of brokers designated as controllers.
¦One indicates that the broker is the controller for the cluster.

¦Leader election rate and time in milliseconds
m¦kafka.controller:type=ControllerStats,name=LeaderElectionRateAndTimeMs
¦The rate at which new leader replicas are elected.
¦Zero

|===

.Yammer metrics

Metrics that express a rate or unit of time are provided as Yammer metrics. The class name of an MBean that uses Yammer metrics is prefixed with `com.yammer.metrics`. 

Yammer rate metrics have the following attributes for monitoring requests:

* Count
* EventType (Bytes)
* FifteenMinuteRate
* RateUnit (Seconds)
* MeanRate
* OneMinuteRate
* FiveMinuteRate

Yammer time metrics have the following attributes for monitoring requests:

* Max
* Min
* Mean
* StdDev

> * 75/95/98/99/999^th^ Percentile

Reviewer: ```suggestion
* 75/95/98/99/99.9^th^ Percentile
```

// Module included in the following assemblies:
//
// assembly-monitoring.adoc

[id='con-important-broker-metrics-{context}']

= Important Kafka broker metrics

> Kafka provides many MBeans for monitoring the performance of the brokers in your Kafka cluster. In the following tables, these broker-level MBeans are organized into server, network, logging, and controller metrics. The metrics apply to an individual broker rather than the entire cluster.

Author: IMO users interested in monitoring will know about MBeans. 

I've added a new sentence to the assembly in order to introduce the term earlier on.

"Like other Java applications, Kafka provides this management information through managed beans, or MBeans."

> *Kafka server metrics*

Author: I bolded this because AsciiDoctor converted the heading into a table caption. I'll try again.

> *Kafka server metrics*

Author: Changed the markup to a sub-heading.

`. Kafka server metrics`

The following table shows a selection of metrics that report information about logging.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected Value

¦Log flush rate and time in milliseconds
m¦kafka.log:type=LogFlushStats,name=LogFlushRateAndTimeMs
¦The rate at which log data is written to disk, in milliseconds.
¦N/A

¦Offline log directory count
m¦kafka.log:type=LogManager,name=OfflineLogDirectoryCount
¦The number of offline log directories (for example, after a hardware failure).
¦Zero

|===

*Kafka controller metrics*

The following table shows a selection of metrics that report information about the controller of the cluster.

[cols="4*",options="header",stripes="none",separator=¦]
|===

¦Metric
¦MBean
¦Description
¦Expected Value

¦Active controller count
m¦kafka.controller:type=KafkaController,name=ActiveControllerCount
¦The number of brokers designated as controllers.
¦One indicates that the broker is the controller for the cluster.

¦Leader election rate and time in milliseconds
m¦kafka.controller:type=ControllerStats,name=LeaderElectionRateAndTimeMs
¦The rate at which new leader replicas are elected.
¦Zero

|===

> .Yammer metrics

Author: Changed the markup to `== Yammer metrics`.

|===

> == Yammer metrics

Reviewer 2: You made this one a heading, but the previous two labels/titles  I think the previous 2 should be headings as well.
